File: cli/src/commands/query.js
import path from 'path';
import { loadGraph } from '../graph.js';
import { querySymbol, queryModule, queryDependants, queryDependencies } from '../query.js';

/**
 * Register the `query` command on the given Commander program.
 *
 * @param {import('commander').Command} program - The Commander program instance.
 */
export function registerQueryCommand(program) {
  program
    .command('query <symbol>')
    .description('Query the code graph for a symbol or module')
    .option('--type <type>', 'Filter by type: function, class, or type')
    .option('--dir <dir>', 'Project directory', '.')
    .action(async (symbol, options) => {
      const rootDir = path.resolve(options.dir);
      const outputDir = path.join(rootDir, '.codemap');

      let graph;
      try {
        graph = await loadGraph(outputDir);
      } catch {
        console.error('No code graph found. Run "codegraph scan" first.');
        process.exit(1);
      }

      // First try as module name
      const moduleResult = queryModule(graph, symbol);
      if (moduleResult) {
        console.log(`Module: ${moduleResult.name}`);
        console.log(`  Files: ${moduleResult.files.join(', ')}`);
        console.log(`  Depends on: ${moduleResult.dependsOn.length > 0 ? moduleResult.dependsOn.join(', ') : '(none)'}`);
        console.log(`  Depended by: ${moduleResult.dependedBy.length > 0 ? moduleResult.dependedBy.join(', ') : '(none)'}`);

        const deps = queryDependencies(graph, symbol);
        const dependants = queryDependants(graph, symbol);
        console.log(`  Dependencies: ${deps.length}`);
        console.log(`  Dependants: ${dependants.length}`);
        return;
      }

      // Then try as symbol name
      const symbolResults = querySymbol(graph, symbol, {
        type: options.type || undefined,
      });

      if (symbolResults.length === 0) {
        console.log(`No results found for "${symbol}".`);
        return;
      }

      console.log(`Found ${symbolResults.length} result(s) for "${symbol}":\n`);
      for (const result of symbolResults) {
        console.log(`  [${result.kind}] ${result.name}`);
        if (result.signature) {
          console.log(`    Signature: ${result.signature}`);
        }
        console.log(`    File: ${result.file}`);
        console.log(`    Module: ${result.module}`);
        console.log(`    Lines: ${result.lines.start}-${result.lines.end}`);
        if (result.calledBy.length > 0) {
          console.log(`    Called by: ${result.calledBy.join(', ')}`);
        }
      }
    });
}

File: rust-cli/src/commands/query.rs
use clap::Args;
use std::path::PathBuf;

#[derive(Args)]
pub struct QueryArgs {
    /// Symbol or module name to query
    pub symbol: String,
    /// Filter by type: function, class, or type
    #[arg(long)]
    pub r#type: Option<String>,
    /// Project directory
    #[arg(long, default_value = ".")]
    pub dir: String,
    /// Query a module instead of a symbol
    #[arg(long)]
    pub module: bool,
}

pub fn run(args: QueryArgs) {
    let root = PathBuf::from(&args.dir);
    let root = match root.canonicalize() {
        Ok(p) => p,
        Err(e) => {
            eprintln!("Error: cannot resolve directory '{}': {}", args.dir, e);
            std::process::exit(1);
        }
    };

    let output_dir = root.join(".codemap");
    let graph = match crate::graph::load_graph(&output_dir) {
        Ok(g) => g,
        Err(e) => {
            eprintln!(
                "Error: failed to load code graph from '{}/.codemap/': {}",
                root.display(),
                e
            );
            eprintln!("Hint: run 'codegraph scan {}' first.", root.display());
            std::process::exit(1);
        }
    };

    if args.module {
        // 模块查询模式
        match crate::query::query_module(&graph, &args.symbol) {
            Some(result) => println!("{}", crate::query::format_module_result(&result)),
            None => {
                eprintln!("Module '{}' not found.", args.symbol);
                // 列出可用模块
                let mut mods: Vec<&str> = graph.modules.keys().map(|s| s.as_str()).collect();
                mods.sort();
                if !mods.is_empty() {
                    eprintln!("Available modules: {}", mods.join(", "));
                }
                std::process::exit(1);
            }
        }
    } else {
        // 符号查询模式
        let opts = crate::query::QueryOptions {
            type_filter: args.r#type.clone(),
        };
        let results = crate::query::query_symbol(&graph, &args.symbol, &opts);
        println!("{}", crate::query::format_symbol_results(&results));
    }
}
